proposals/0012-introduce-reactNativeMetadata.md

```diff
@@ -0,0 +1,329 @@
+* **verifiability**: the manifest can be easily verified & linted by tooling to ensure that app & library are not using an incompatible set of compatibilites. Similarly, the JSON schema we publish can be used to validate the `reactNativeManifest` section.
+
+## Use Cases
+
+This section lists the use case we envision for the `reactNativeManifest` section.
+
+### Declaring the type of a project
+
+For the sake of this RFC, we consider a React Native project as either an app or a library.
+While an app project can contain native modules and components, it's **not** possible for a app to "consume" another app as a library as this would break the native build system invariants (i.e. a single `AndroidManifest.xml` and `Info.plist` file per app, etc.). If an app projects wishes to distribute some of its native modules or components, they need to be extracted into a library project.
+
+We're proposing to declare if a project is either a library or an app by adding the `type` field (as `"app"` or `"library"`) to the `reactNativeManifest` section of the `package.json` file.
+
+```json
+{
+  "reactNativeManifest": {
+    "type": "library"
+  }
+}
+```
+
+### New Architecture support
+
+One of the primary driver of this proposal is the [New React Native Architecture support](https://reactnative.dev/docs/the-new-architecture/landing-page). Specifically we believe that the ecosystem is currently affected by:
+* Lack of a declarative way to define if a library is compatible with New Architecture
+  * As described in the [rationale](#rationale), tooling needs to infer this information from the presence of the `codegenConfig` field in the `package.json` of a library. Libraries might decide to don't use codegen and still be compatible with New Architecture, so this method is not exhaustive.
+  * The alternative at this stage is to inspect the code and check the API callsites if they are New Architecture or Old Architecture compatible.
+* Lack of a declarative way to enable New Architecture on both platforms
+  * Currently New Architecture is enabled by a Gradle Property `newArchEnabled` on Android in the `gradle.properties` file and by invokign `RCT_NEW_ARCH_ENABLED pod install` on iOS.
+  * As of today, there is no way to enable New Architecture for an app project for both platforms
+  * Moreover, the file vs environment variable leads to a scenario where you can't know if a app supports New Architecture or not by inspecting the code (as New Architecture support is known at  `pod install` time and is not codified in the codebase)
+
+Therefore we propose to add the `newArch` section to the `reactNativeManifest.capabilities` of **both apps and libraries** with the following semantic:
+* For Apps: `newArch.enabled==true` implies that the app wants to use the New Architecture.
+* For Libraries: `newArch.enabled==true` implies that the library is compatible with the New Architecture.
```

Comment:
Is the thinking then that there's no way to express one platform only supporting newArch? And it's not something we want to support?